vllm/model_executor/layers/quantization/modelopt.py

```diff
@@ -721,6 +748,80 @@ def __init__(self, quant_config: ModelOptNvFp4Config):
                 raise ValueError("Current platform does not support NVFP4"
                                  " quantization. Please use Blackwell and"
                                  " above.")
+        from vllm.model_executor.layers.fused_moe.cutlass_moe import (
+            cutlass_moe_fp4)
+
+        self.fused_experts = cutlass_moe_fp4
+
+    @property
+    def load_up_proj_weight_first(self) -> bool:
+        # FlashInfer CUTLASS kernel assumes [Up, Gate] Proj as W13
+        return self.allow_flashinfer_cutlass
+
+    def select_experts_impl(self, moe_parallel_config):
+        if not self.allow_flashinfer_cutlass:
+            return
+
+        logger.debug("FlashInferExperts")
+        # default to TP/EP case only
+
+        experts_kwargs = {
+            "use_nvfp4_w4a4": True,
+            "use_dp": moe_parallel_config.dp_size > 1,
+        }
+        # if not moe_parallel_config.dp_size > 1 and moe_parallel_config.use_ep:
+        experts_kwargs["ep_rank"] = moe_parallel_config.ep_rank
+        experts_kwargs["ep_size"] = moe_parallel_config.ep_size
+        experts_kwargs["tp_rank"] = moe_parallel_config.tp_rank
+        experts_kwargs["tp_size"] = moe_parallel_config.tp_size
+        experts = FlashInferExperts(**experts_kwargs)
+        self.fused_experts = mk.FusedMoEModularKernel(
+            FlashInferCutlassMoEPrepareAndFinalize(
+                quant_dtype=torch.uint8,
+                #meaning 2x e2m1 packed in one, kernel requirement
+            ),
+            experts,
+        )
+
+    @property
+    def load_up_proj_weight_first(self) -> bool:
+        # FlashInfer CUTLASS kernel assumes [Up, Gate] Proj as W13
+        if self.allow_flashinfer_cutlass:
+            return True
+        return False
+
+    # This method update self.fused_experts
+    # only prepare_finalize is not None call select_gemm_impl
+    # so when native cutlass fp4, fused_expert is in fuse_moe.py fused_expert
+    # when it's not called(TP case), we still have 2 kernels to use.
+    def select_gemm_impl(self, prepare_finalize, moe):
+
+        assert moe is not None
+        assert prepare_finalize is not None
+        experts = None
+        all2all_manager = get_ep_group().device_communicator.all2all_manager
+        assert all2all_manager is not None
+        if self.allow_flashinfer_cutlass:
+            from vllm.model_executor.layers.fused_moe.flashinfer_cutlass_moe import (
+                FlashInferExperts)
+            logger.debug("FlashInferExperts %s", moe)
```

Comment:
Could this be a debug_once or combined with the log above?
```suggestion
            logger.debug_once("Using FlashInferExperts %s", moe)
```